backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/v2/invoice/review/impl/JsonInvoiceReviewService.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * #%L
+ * de.metas.business.rest-api-impl
+ * %%
+ * Copyright (C) 2023 metas GmbH
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+package de.metas.rest_api.v2.invoice.review.impl;
+
+import de.metas.RestUtils;
+import de.metas.common.rest_api.common.JsonMetasfreshId;
+import de.metas.common.util.CoalesceUtil;
+import de.metas.invoice.InvoiceId;
+import de.metas.invoice.InvoiceQuery;
+import de.metas.invoice.review.InvoiceReview;
+import de.metas.invoice.review.InvoiceReviewId;
+import de.metas.invoice.service.IInvoiceDAO;
+import de.metas.invoice.service.impl.InvoiceReviewRepository;
+import de.metas.organization.OrgId;
+import de.metas.rest_api.utils.IdentifierString;
+import de.metas.rest_api.v2.invoice.JsonCreateInvoiceReviewResponse;
+import de.metas.rest_api.v2.invoice.JsonCreateInvoiceReviewResponseResult;
+import de.metas.rest_api.v2.invoice.JsonInvoiceReviewUpsertItem;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.adempiere.exceptions.AdempiereException;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+
+@Service
+public class JsonInvoiceReviewService
+{
+	private final InvoiceReviewRepository invoiceReviewRepository;
+	private final IInvoiceDAO invoiceDAO = Services.get(IInvoiceDAO.class);
+
+	public JsonInvoiceReviewService(@NonNull final InvoiceReviewRepository invoiceReviewRepository)
+	{
+		this.invoiceReviewRepository = invoiceReviewRepository;
+	}
+
+	public Optional<JsonCreateInvoiceReviewResponse> upsert(@NonNull final JsonInvoiceReviewUpsertItem jsonInvoiceReviewUpsertItem)
+	{
+		final OrgId orgId = getOrgId(jsonInvoiceReviewUpsertItem.getOrgCode());
+		final InvoiceQuery invoiceQuery = createInvoiceQueryOrNull(jsonInvoiceReviewUpsertItem, orgId);
+		if (invoiceQuery == null)
+		{
+			return Optional.empty();
+		}
+		final Optional<InvoiceId> invoiceId = invoiceDAO.retrieveIdByInvoiceQuery(invoiceQuery);
```

Comment:
matter of taste: generally I would avoid having `Optional` fields or local variables,
because you will pollute the whole code if `.get()` calls.

approach 1. Null approach
```java
InvoiceId invoiceId = invoiceDAO.retrieveIdByInvoiceQuery(invoiceQuery).orElse(null);
if(invoiceId == null) return Optoinal.empty();
```

approach 2. `Optional.map` approach
```java
return invoiceDAO.retrieveIdByInvoiceQuery(invoiceQuery)
.map(invoiceId -> upsert(invoiceId, jsonInvoiceReviewUpsertItem);
```

and you will refactor the code below in a new method called upsert (or a better descriptive name).

----

Depends on this case. In this case, I would prefer approach 1 but in some other cases, I would go with approach 2 when it feels more natural.

PS: Also if u leave it like it is, is also fine.

Reply:
went with approach 2 since the logic below simplified a bit.